### ISSUE TITLE
Child process PID is incorrect when using crystal's official docker image

### ISSUE THREAD

**Issue by samueleaton** [crystal-lang/crystal#3431]
I don't know if this is a difference with Mac vs Linux in the way that processes are generated, but there is an inconsistency.

I have created a repo to highlight my problem for your convenience here: https://github.com/samueleaton/crystal_child_process_bug

### Files

```ruby
# parent.cr
puts "parent pid from parent: #{Process.pid}"
child_process = Process.new("./child", shell: true, output: true, error: true)
puts "child pid from parent: #{child_process.pid}"
sleep 2
```

```ruby
# child.cr
puts "\n"
puts "parent pid from child: #{Process.ppid}"
puts "child pid from child: #{Process.pid}"
```

```bash
# Dockerfile
FROM crystallang/crystal
MAINTAINER sameaton11@gmail.com
WORKDIR /app
ADD . ./
RUN crystal build parent.cr
RUN crystal build child.cr
CMD ./parent
```

### Description

👍  When running without docker, the PIDs are as expected for the parent and child. (`x` and `y` would be replace by respective PIDs) 

```
parent pid from parent: x
child pid from parent: y

parent pid from child: x
child pid from child: y
```

👎  However, when running in Docker, there is an inconsistency with the PIDs. 

```
parent pid from parent: x
child pid from parent: y

parent pid from child: y
child pid from child: z
```

This makes me believe that the parent process generates a "middleman process" that in turn generate the child process.

The parent and child have the same process group id either way.

**Comment by samueleaton** [comment]
I have discovered that this bug only occurs when `shell: true`.

**Comment by RX14** [comment]
It does seems that a shell is there as an intermediary.

![](https://aww.moe/9pnz48.png)

Why does this work outside the container? The generated shell command doesn't seem to use `exec`. Could we use `exec` in `process.prepare_argv`? Maybe we should simply leave it and not guarantee any behaviour.

**Comment by kirbyfan64** [comment]
Crystal's docker image is based on Debian, right?

Bash will automatically `exec` a process if it comes last, but most other shells don't. My guess is that, if you're running this locally and your current shell is Bash, then the process is automatically `exec`'d.

BUT, on Debian, the default shell is `dash`, which is based on Posix sh. I don't believe this one automatically `exec`s, which would explain the discrepancy.

Most people wouldn't notice the difference because most people just use Bash anyhow.

Regardless, though, this isn't really a bug. Maybe it could be documented a bit better, but it's working correctly.

@RX14 Using `exec` would break running Bash builtins or aliases.

**Comment by RX14** [comment]
@kirbyfan64 I didn't think of that, thanks.

**Comment by mverzilli** [comment]
Thanks for the insights @kirbyfan64 🙇 . Would you help me decide how to follow up on this? I guess for now we could add a wiki page about working with Docker to document this and future caveats, then close this. WDYT?

**Comment by kirbyfan64** [comment]
TBH this isn't really a Docker thing; it would happen any time your login shell is something more Posix-y than Bash. Maybe it could just be a side note in the `Process` docs?

**Comment by RX14** [comment]
While in general I don't think we can document every detail of POSIX in the crystal docs, this detail may have an affect on the killing of child processes, so perhaps it would be a good idea to document.

**Comment by jhass** [comment]
I agree  it's not a bug, but why wouldn't we improve things more by making sure the  (last) subcommand is always `exec'ed if possible?  That would make the behavior consistent across shell's.

You would hope though that `bash` would turn off such special behavior if invoked as `/bin/sh`  :/
